app/javascript/role_manager.js

```diff
@@ -0,0 +1,222 @@
+        if (target_role == $(column).data('roleManagerRoleValue')) {
+          $(column).addClass('hide')
+        }
+      });
+    }
+  }
+
+  toggleAdmin(e) {
+    const target = $(e.currentTarget)
+    const input = $(target).find('input')
+    // toggle the visibility of the associated administrative items
+    if ($(input).val() == 'show') {
+      $(this.administrativeInputTargets).removeClass('hide')
+    } else {
+      $(this.administrativeInputTargets).addClass('hide')
+    }
+  }
+
+  searchPermissions(e) {
+    // if we have more than three characters
+    // 1. Expand all sections
+    // 2. hide any permission where the search string doesn't exit in the text
+    const target = $(e.currentTarget)
+    const search_string = target.val().toLowerCase()
+    if (search_string.length > 2) {
+      this.permissionCategoryTargets.forEach((section) => {
+        $(section).siblings('.panel-collapse').collapse('show')
+      });
+      this.individualPermissionTargets.forEach((permission) => {
+        const wrapper = $(permission).closest('.form-check')
+        const permission_text = wrapper.text().toLowerCase()
+        if (permission_text.indexOf(search_string) == -1) {
+          wrapper.addClass('hide')
+        } else {
+          wrapper.removeClass('hide')
+        }
+      });
+    } else {
+      // reset
+      this.permissionCategoryTargets.forEach((section) => {
+        $(section).siblings('.panel-collapse').collapse('hide')
+      });
+      this.individualPermissionTargets.forEach((permission) => {
+        $(permission).closest('.form-check').removeClass('hide')
+      })
+    }
+  }
+
+  // find all inputs in the form, store key value pairs for later comparision
+  setInitialState() {
+    this.initialState = {}
+    this.setState(this.initialState)
+  }
+
+  setState(variable) {
+    const inputs = $(this.roleColumnTargets).find('input')
+    $(inputs).each((i) => {
+      const input = inputs[i]
+      // console.log(inputs[i])
+      if ($(input).is(':checked')) {
+        variable[$(input).attr('name')] = '1'
+      } else {
+        variable[$(input).attr('name')] = '0'
+      }
+    })
+  }
+
+  // Collect all form fields, determine if any have changed from their initial state
+  // make note by enabling the save button and count of changes if there are any
+  updateState(e) {
+    this.currentState = {}
+    this.setState(this.currentState)
+    this.updateUi()
+  }
+
+  updateUi() {
+    let changed = 0
+    $.each(this.currentState, (key, value) => {
+      if (this.initialState[key] != value) {
+        changed += 1
+      }
+    })
+    if (changed == 0) {
+      $(this.changeCountTarget).text('')
+      $(this.changeButtonTarget).addClass('hide')
+    } else if (changed == 1) {
+      $(this.changeCountTarget).text(`${changed} change pending`)
+      $(this.changeButtonTarget).removeClass('hide')
+    } else {
+      $(this.changeCountTarget).text(`${changed} changes pending`)
```

Comment:
This is nice, but I am a little sad to lose the yellow-check that showed me which changes I had made that weren't saved yet (at least for additions). Here, I can add a bunch of perms, but I can't see at a glance which ones I have added before hitting save. I usually will review that just to make sure. So.. not sure how tricky that is, but I would prefer to have some treatment showing me _which_ perms have an unsaved change. (Take or leave maybe scope-creep for this pr!)

Reply:
restored!